File: backend/src/test/java/org/globe42/web/persons/ParticipationControllerMvcTest.java
package org.globe42.web.persons;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.PersonDao;
import org.globe42.domain.ActivityType;
import org.globe42.domain.Participation;
import org.globe42.domain.Person;
import org.globe42.test.GlobeMvcTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC tests for {@link ParticipationController}
 * @author JB Nizet
 */
@RunWith(SpringRunner.class)
@GlobeMvcTest(ParticipationController.class)
public class ParticipationControllerMvcTest {
    @MockBean
    private PersonDao mockPersonDao;

    @Autowired
    private MockMvc mvc;

    @Autowired
    private ObjectMapper objectMapper;

    private Person person;
    private Participation mealParticipation;

    @Before
    public void prepare() {
        person = new Person(42L);
        mealParticipation = new Participation(34L);
        mealParticipation.setActivityType(ActivityType.MEAL);
        person.addParticipation(mealParticipation);
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));
    }

    @Test
    public void shouldList() throws Exception {
        mvc.perform(get("/api/persons/{personId}/participations", person.getId()))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(34));
    }

    @Test
    public void shouldCreate() throws Exception {
        ParticipationCommandDTO command = new ParticipationCommandDTO(ActivityType.SOCIAL_MEDIATION);
        mvc.perform(post("/api/persons/{personId}/participations", person.getId())
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.activityType").value(ActivityType.SOCIAL_MEDIATION.name()));
    }

    @Test
    public void shouldDelete() throws Exception {
        mvc.perform(delete("/api/persons/{personId}/participations/{participationId}", person.getId(), mealParticipation.getId()))
           .andExpect(status().isNoContent());
    }
}

File: backend/src/test/kotlin/org/globe42/web/persons/ParticipationControllerMvcTest.kt
package org.globe42.web.persons

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.globe42.dao.PersonDao
import org.globe42.domain.ActivityType
import org.globe42.domain.Gender
import org.globe42.domain.Participation
import org.globe42.domain.Person
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import java.util.*

/**
 * MVC tests for [ParticipationController]
 * @author JB Nizet
 */
@GlobeMvcTest(ParticipationController::class)
class ParticipationControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {
    @MockkBean
    private lateinit var mockPersonDao: PersonDao

    private lateinit var person: Person
    private lateinit var mealParticipation: Participation

    @BeforeEach
    fun prepare() {
        mealParticipation = Participation(34L)
        mealParticipation.activityType = ActivityType.MEAL
        person = Person(42L, "John", "Doe", Gender.MALE)
        person.addParticipation(mealParticipation)
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person
    }

    @Test
    fun `should list`() {
        mvc.get("/api/persons/{personId}/participations", person.id).andExpect {
            status { isOk }
            jsonValue("$[0].id", 34)
        }
    }

    @Test
    fun `should create`() {
        val command = ParticipationCommandDTO(ActivityType.SOCIAL_MEDIATION)
        every { mockPersonDao.flush() } answers {
            person.getParticipations().find { it.activityType == ActivityType.SOCIAL_MEDIATION }?.let { it.id = 345L }
            Unit
        }
        mvc.post("/api/persons/{personId}/participations", person.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isCreated }
            jsonValue("$.activityType", ActivityType.SOCIAL_MEDIATION.name)
        }
    }

    @Test
    fun `should delete`() {
        mvc.delete("/api/persons/{personId}/participations/{participationId}", person.id, mealParticipation.id)
            .andExpect {
                status { isNoContent }
            }
    }
}
